presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/MetadataQueryOptimizer.java

```diff
@@ -151,31 +166,111 @@ public PlanNode visitAggregation(AggregationNode node, RewriteContext<Void> cont
                 return context.defaultRewrite(node);
             }
 
-            ImmutableList.Builder<List<RowExpression>> rowsBuilder = ImmutableList.builder();
-            for (TupleDomain<ColumnHandle> domain : predicates.getPredicates()) {
-                if (!domain.isNone()) {
-                    Map<ColumnHandle, NullableValue> entries = TupleDomain.extractFixedValues(domain).get();
+            if (!isReducible(node)) {
+                ImmutableList.Builder<List<RowExpression>> rowsBuilder = ImmutableList.builder();
+                for (TupleDomain<ColumnHandle> domain : predicates.getPredicates()) {
+                    if (!domain.isNone()) {
+                        Map<ColumnHandle, NullableValue> entries = TupleDomain.extractFixedValues(domain).get();
+
+                        ImmutableList.Builder<RowExpression> rowBuilder = ImmutableList.builder();
+                        // for each input column, add a literal expression using the entry value
+                        for (VariableReferenceExpression input : inputs) {
+                            ColumnHandle column = columns.get(input);
+                            NullableValue value = entries.get(column);
+                            if (value == null) {
+                                // partition key does not have a single value, so bail out to be safe
+                                return context.defaultRewrite(node);
+                            }
+                            else {
+                                rowBuilder.add(constant(value.getValue(), input.getType()));
+                            }
+                        }
+                        rowsBuilder.add(rowBuilder.build());
+                    }
+                }
 
-                    ImmutableList.Builder<RowExpression> rowBuilder = ImmutableList.builder();
-                    // for each input column, add a literal expression using the entry value
-                    for (VariableReferenceExpression input : inputs) {
-                        ColumnHandle column = columns.get(input);
+                // replace the tablescan node with a values node
+                return SimplePlanRewriter.rewriteWith(new Replacer(new ValuesNode(idAllocator.getNextId(), inputs, rowsBuilder.build())), node);
+            }
+
+            // Fold min/max aggregations to a constant value
+            ImmutableList.Builder<RowExpression> scalarsBuilder = ImmutableList.builder();
+            for (int i = 0; i < inputs.size(); i++) {
+                ImmutableList.Builder<RowExpression> arguments = ImmutableList.builder();
+                ColumnHandle column = columns.get(inputs.get(i));
+                // for each input column, add a literal expression using the entry value
+                for (TupleDomain<ColumnHandle> domain : predicates.getPredicates()) {
+                    if (!domain.isNone()) {
```

Comment:
```java
if (domain.isNone()) {
    continue;
}

Reply:
This is following the existing code, but I agree with the proposed change. Also changed the original code as well.